(indirect-methods-exercises-section)=

# Indirect methods exercises

Answer the following exercises based on the content from this chapter. The solutions can be found in the [appendices](indirect-methods-solutions-section).

````{exercise}
:label: indirect-methods-ex-jacobi

Using a pen and calculator, calculate the first 2 iterations of the Jacobi method for solving the system of linear equations below. Use starting values of $x_i^{(0)} = 0 $ and work to 4 decimal places.

$$ \begin{align*}
    4x_1 +x_2 -x_3 +x_4 &=14,\\
    x_1 +4x_2 -x_3 -x_4 &=10,\\
    -x_1 -x_2 +5x_3 +x_4 &=-15,\\
    x_1 -x_2 +x_3 +3x_4 &=3.
\end{align*} $$

```{dropdown} Solution
|  $k$  |   $x_1$   |   $x_2$   |   $x_3$   |   $x_4$   |
|:---:|:--------:|:--------:|:--------:|:--------:|
|   0 |   0.0000 |   0.0000 |   0.0000 |   0.0000 |
|   1 |   3.5000 |   2.5000 |  -3.0000 |   1.0000 |
|   2 |   1.8750 |   1.1250 |  -2.0000 |   1.6667 |
```
````


In [1]:
import numpy as np

def jacobi(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    print("|  k  |", end="")
    for i in range(n):
        print(f"   x({i+1})   |", end="")
    print("\n|:---:|" + ":--------:|" * n)
    print(f"| {0:3d} |", end="")
    for i in range(n):
        print(f" {x[i]:8.4f} |", end="")
    print()

    for k in range(maxiter):
        xold = np.copy(x)
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * xold[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break

        
        print(f"| {k+1:3d} |", end="")
        for i in range(n):
            print(f" {x[i]:8.4f} |", end="")
        print()
    
    return x


 # Define system
A = np.array([[4, 1, -1, 1], [1, 4, -1, -1], [-1, -1, 5, 1], [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Solve system
x = jacobi(A, b, tol=1e-4)
print(f"x = {x}")

|  k  |   x(1)   |   x(2)   |   x(3)   |   x(4)   |
|:---:|:--------:|:--------:|:--------:|:--------:|
|   0 |   0.0000 |   0.0000 |   0.0000 |   0.0000 |
|   1 |   3.5000 |   2.5000 |  -3.0000 |   1.0000 |
|   2 |   1.8750 |   1.1250 |  -2.0000 |   1.6667 |
|   3 |   2.3021 |   1.9479 |  -2.7333 |   1.4167 |
|   4 |   1.9755 |   1.5953 |  -2.4333 |   1.7931 |
|   5 |   2.0446 |   1.8461 |  -2.6444 |   1.6844 |
|   6 |   1.9563 |   1.7488 |  -2.5587 |   1.8153 |
|   7 |   1.9693 |   1.8251 |  -2.6220 |   1.7838 |
|   8 |   1.9423 |   1.7981 |  -2.5979 |   1.8259 |
|   9 |   1.9445 |   1.8214 |  -2.6171 |   1.8179 |
|  10 |   1.9359 |   1.8141 |  -2.6104 |   1.8313 |
|  11 |   1.9360 |   1.8213 |  -2.6163 |   1.8295 |
|  12 |   1.9332 |   1.8193 |  -2.6144 |   1.8338 |
|  13 |   1.9331 |   1.8215 |  -2.6163 |   1.8335 |
|  14 |   1.9322 |   1.8210 |  -2.6158 |   1.8349 |
|  15 |   1.9321 |   1.8217 |  -2.6163 |   1.8349 |
|  16 |   1.9318 |   1.8216 |  -2.6162 |   1.8353 |
|  17 |   1.


````{exercise} 
:label: indirect-methods-ex-gauss-seidel

Repeat {ref}`indirect-methods-ex-jacobi` using the Gauss-Seidel method.

```{dropdown} Solution
|  k  |   x(1)   |   x(2)   |   x(3)   |   x(4)   |
|:---:|:--------:|:--------:|:--------:|:--------:|
|   0 |   0.0000 |   0.0000 |   0.0000 |   0.0000 |
|   1 |   3.5000 |   1.6250 |  -1.9750 |   1.0333 |
|   2 |   2.3417 |   1.6792 |  -2.4025 |   1.5800 |
```
````


In [2]:
import numpy as np

def gauss_seidel(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    print("|  k  |", end="")
    for i in range(n):
        print(f"   x({i+1})   |", end="")
    print("\n|:---:|" + ":--------:|" * n)
    print(f"| {0:3d} |", end="")
    for i in range(n):
        print(f" {x[i]:8.4f} |", end="")
    print()

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break

        
        print(f"| {k+1:3d} |", end="")
        for i in range(n):
            print(f" {x[i]:8.4f} |", end="")
        print()
    
    return x


 # Define system
A = np.array([[4, 1, -1, 1], [1, 4, -1, -1], [-1, -1, 5, 1], [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Solve system
x = gauss_seidel(A, b, tol=1e-4)
print(f"x = {x}")

|  k  |   x(1)   |   x(2)   |   x(3)   |   x(4)   |
|:---:|:--------:|:--------:|:--------:|:--------:|
|   0 |   0.0000 |   0.0000 |   0.0000 |   0.0000 |
|   1 |   3.5000 |   1.6250 |  -1.9750 |   1.0333 |
|   2 |   2.3417 |   1.6792 |  -2.4025 |   1.5800 |
|   3 |   2.0846 |   1.7732 |  -2.5444 |   1.7444 |
|   4 |   1.9845 |   1.8039 |  -2.5912 |   1.8035 |
|   5 |   1.9504 |   1.8155 |  -2.6075 |   1.8242 |
|   6 |   1.9382 |   1.8196 |  -2.6133 |   1.8316 |
|   7 |   1.9339 |   1.8211 |  -2.6153 |   1.8342 |
|   8 |   1.9323 |   1.8216 |  -2.6160 |   1.8351 |
|   9 |   1.9318 |   1.8218 |  -2.6163 |   1.8354 |
|  10 |   1.9316 |   1.8219 |  -2.6164 |   1.8356 |
x = [ 1.93154463  1.82190482 -2.61642052  1.83559357]



````{exercise}
:label: indirect-methods-ex-sor

Repeat {ref}`indirect-methods-ex-jacobi` using the SOR method using the optimum value for the relaxation parameter.

```{dropdown} Solution

$\omega = 1.1136$

|  k  |   x(1)   |   x(2)   |   x(3)   |   x(4)   |
|:---:|:--------:|:--------:|:--------:|:--------:|
|   0 |   0.0000 |   0.0000 |   0.0000 |   0.0000 |
|   1 |   3.8977 |   1.6989 |  -2.0944 |   1.0749 |
|   2 |   2.0994 |   1.7227 |  -2.4910 |   1.7763 |

````


In [3]:
import numpy as np

def sor(A, b, omega, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    print("|  k  |", end="")
    for i in range(n):
        print(f"   x({i+1})   |", end="")
    print("\n|:---:|" + ":--------:|" * n)
    print(f"| {0:3d} |", end="")
    for i in range(n):
        print(f" {x[i]:8.4f} |", end="")
    print()

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (1 - omega) * x[i] + omega * (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break

        
        print(f"| {k+1:3d} |", end="")
        for i in range(n):
            print(f" {x[i]:8.4f} |", end="")
        print()
    
    return x


 # Define system
A = np.array([[4, 1, -1, 1], [1, 4, -1, -1], [-1, -1, 5, 1], [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Calculate omega
D = np.diag(np.diag(A))
U = -np.tril(A, -1)
L = -np.triu(A, 1)
TJ = -np.dot(np.linalg.inv(D), L + U)
rho = max(abs(np.linalg.eigvals(TJ)))
omega =  1 + (rho / (1 + np.sqrt(1 - rho ** 2))) ** 2
print(f"omega = {omega:0.4f}\n")

# Solve system
x = sor(A, b, omega, tol=1e-4)
print(f"x = {x}")

omega = 1.1136

|  k  |   x(1)   |   x(2)   |   x(3)   |   x(4)   |
|:---:|:--------:|:--------:|:--------:|:--------:|
|   0 |   0.0000 |   0.0000 |   0.0000 |   0.0000 |
|   1 |   3.8977 |   1.6989 |  -2.0944 |   1.0749 |
|   2 |   2.0994 |   1.7227 |  -2.4910 |   1.7763 |
|   3 |   1.9915 |   1.8349 |  -2.6012 |   1.8193 |
|   4 |   1.9299 |   1.8206 |  -2.6152 |   1.8371 |
|   5 |   1.9320 |   1.8227 |  -2.6166 |   1.8356 |
|   6 |   1.9312 |   1.8219 |  -2.6165 |   1.8357 |
|   7 |   1.9315 |   1.8219 |  -2.6165 |   1.8356 |
x = [ 1.9314956   1.82191405 -2.61644014  1.83561973]



````{exercise}
:label: indirect-methods-ex-convergence

Calculate the spectral radii for the three methods used to compute the solution to the system from {ref}`indirect-methods-ex-jacobi`. Which of the three methods would you expect to converge to a solution the fastest?

```{dropdown} Solution
The spectral radii of the iteration matrices for the three methods are

$$ \begin{align*}
  \rho(T_J) &= 0.6054, \\
  \rho(T_{GS}) &= 0.3553, \\
  \rho(T_{SOR}) &= 0.1968.
\end{align*} $$

So the SOR method should converge the fastest.
```

````


In [4]:
# Calculate omega
D = np.diag(np.diag(A))
U = -np.tril(A, -1)
L = -np.triu(A, 1)

TJ = -np.dot(np.linalg.inv(D), L + U)
TGS = -np.dot(np.linalg.inv(L + D), U)
TSOR = np.dot(np.linalg.inv(D + omega * L), (1 - omega) * D - omega * U)

rho_TJ = max(abs(np.linalg.eigvals(TJ)))
rho_TGS = max(abs(np.linalg.eigvals(TGS)))
rho_TSOR = max(abs(np.linalg.eigvals(TSOR)))

print(f"rho_TJ = {rho_TJ:0.4f}")
print(f"rho_TGS = {rho_TGS:0.4f}")
print(f"rho_TSOR = {rho_TSOR:0.4f}")


rho_TJ = 0.6054
rho_TGS = 0.2353
rho_TSOR = 0.1995



````{exercise}
:label: indirect-methods-ex-code

Write a program to calculate the solution to the system of linear equations from {ref}`indirect-methods-ex-jacobi` using the Jacobi, Gauss-Seidel and SOR methods using a convergence tolerance of $tol=10^{-6}$. How many iterations did each of the three methods take to converge to the solution?

```{dropdown} Solution
$\mathbf{x} = (1.931496, 1.821914, -2.616440, 1.835620)$

- Jacobi: 19 iterations
- Gauss-Seidel: 10 iterations
- SOR: 7 iterations
```
````


In [5]:
import numpy as np

def jacobi(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    for k in range(maxiter):
        xold = np.copy(x)
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * xold[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break
    
    return x, k


def gauss_seidel(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break
    
    return x, k

def sor(A, b, omega, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100

    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (1 - omega) * x[i] + omega * (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break
    
    return x, k


 # Define system
A = np.array([[4, 1, -1, 1], [1, 4, -1, -1], [-1, -1, 5, 1], [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Solve system
x, k_J = jacobi(A, b, tol=1e-4)
x, k_GS = gauss_seidel(A, b, tol=1e-4)
x, k_SOR = sor(A, b, omega, tol=1e-4)
print(f"x = {x}")
print(f"Jacobi: {k_J}")
print(f"Gauss-Seidel: {k_GS}")
print(f"SOR: {k_SOR}")

x = [ 1.9314956   1.82191405 -2.61644014  1.83561973]
Jacobi: 19
Gauss-Seidel: 10
SOR: 7



````{exercise}
:label: indirect-methods-ex-convergence-2

A linear system has the following coefficient matrix. What is the largest value that $\alpha$ can be in order for the Jacobi method to be convergent?

$$ \begin{align*}
    A = \begin{pmatrix}
        2 & 1 \\
        \alpha  & 2
    \end{pmatrix}
\end{align*} $$

```{dropdown} Solution
$\alpha < \sqrt{2}$
```
````

